### PR TITLE
config: add "replace" function [GH-1029]

### LINKED ISSUE #1029: Feature Request: gsub
I've run into a couple situations where I need to format an output string from resourceA as an input to resourceB.

As an example, suppose you want to create an aws_route53_record.  The `name` of the record needs to be built from the public_ip of the attached node.
If the public_ip is "12.34.56.78", the record's `name` has to be "foo-12-34-56-78".

```
resource "aws_route53_record" "foo" {
zone_id = "${var.some_zone_id}"
name = ???  # "foo-${gsub('.', '-', aws_eip.foo.public_ip }"
type = "A"
ttl = "300"
records = ["${aws_eip.foo.public_ip}"]
}
```

The only way I can think to do this now is to write a local-exec provisioner that does string manipulation, writes a json file, and shells out to `awscli`.  Which isn't pretty.

### PR DESCRIPTION
**Pull request by mitchellh**
Fixes #1029 

This adds a `replace` function. From the docs:

```
  * `replace(string, search, replace)` - Does a search and replace on the
      given string. All instances of `search` are replaced with the value
      of `replace`. If `search` is wrapped in forward slashes, it is treated
      as a regular expression. If using a regular expression, `replace`
      can reference subcaptures in the regular expression by using `$n` where
      `n` is the index or name of the subcapture. If using a regular expression,
      the syntax conforms to the [re2 regular expression syntax](https://code.google.com/p/re2/wiki/Syntax).
```

### CODE REVIEW

File: config/interpolate_funcs_test.go [hunk 1]
```diff
@@ -225,7 +258,9 @@ func testFunction(t *testing.T, config testFunctionConfig) {
 		}
 
 		if !reflect.DeepEqual(out, tc.Result) {
-			t.Fatalf("%d: bad: %#v", i, out)
+			t.Fatalf(
+				"%d: bad output for input: %s\n\nOutput: %#v",
+				i, tc.Input, out)
```
Comment (mitchellh) [inline config/interpolate_funcs_test.go]: This one is for you @phinze.

**Comment by phinze** [comment]
LGTM! I can barely keep up with you this morning! :stuck_out_tongue:

**Comment by mitchellh** [comment]
@phinze I've just been doing a bunch of easy things while you toil away at the hard things. ;) One of those "I need easier things" mornings. Haha.

**Pull request merged**